pytorch_lightning/callbacks/stochastic_weight_avg.py

```diff
@@ -280,3 +291,44 @@ def avg_fn(
     ) -> torch.FloatTensor:
         """Adapted from https://github.com/pytorch/pytorch/blob/v1.7.1/torch/optim/swa_utils.py#L95-L97."""
         return averaged_model_parameter + (model_parameter - averaged_model_parameter) / (num_averaged + 1)
+
+    def on_save_checkpoint(
+        self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", checkpoint: Dict[str, Any]
+    ) -> dict:
+        checkpoint_data = {
+            "n_averaged": self.n_averaged,
+            "scheduler_step_count": None if self._swa_scheduler is None else self._swa_scheduler._step_count,
+            "average_model_parameters": self._get_average_model_parameters(trainer),
+        }
+        return checkpoint_data
+
+    def on_load_checkpoint(
+        self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", callback_state: Dict[str, Any]
+    ) -> None:
+        if callback_state:
+            self.n_averaged = callback_state["n_averaged"]
+            self._scheduler_step_count = callback_state["scheduler_step_count"]
+            self._load_average_model_parameters(callback_state["average_model_parameters"])
+        else:
+            rank_zero_warn(
+                f"Checkpoint has no data for the {self.state_key} callback, not initializing the callback state."
+            )
+
+    def _get_average_model_parameters(self, trainer: "pl.Trainer") -> Any:
```

Comment:
Might be better to rely on the training_type_plugin to reload the weights. This won't work with sharded models.

Reply:
I had a go at setting up a unit test with a sharded model following https://pytorch-lightning.readthedocs.io/en/stable/advanced/advanced_gpu.html, setting `strategy='fsdp'` with 2 gpus and defining `configure_sharded_model` on `SwaTestModel`. But the test crashes in `StochasticWeightAveraging.on_fit_start` with ``MisconfigurationException: SWA currently works with 1 `optimizer`.`` When I check the number of optimizers at this point it's actually zero. Does this mean SWA already doesn't work with sharded models or am I doing something wrong?

I'd be happy to change to using the `training_type_plugin` here regardless, but I'm struggling to see how to use it to save and load the weights.

This did also prompt me to test with multiple processes and I found there was a bug in how I handled `n_averaged` as after resuming, each process was using the same instance of `n_averaged`, resulting in it being incremented too many times, so I've fixed that.

Reply:
If I remove the check for 1 optimizer in `on_fit_start`, then by the time we hit `on_train_epoch_start` the number of optimizers is 1 and things work up until calling `update_parameters`, which fails with:
```
        for p_swa, p_model in zip(average_model.parameters(), model.parameters()):
            device = p_swa.device
            p_swa_ = p_swa.detach()
            p_model_ = p_model.detach().to(device)
            src = p_model_ if n_averaged == 0 else avg_fn(p_swa_, p_model_, n_averaged.to(device))
>           p_swa_.copy_(src)
E           RuntimeError: The size of tensor a (32) must match the size of tensor b (528) at non-singleton dimension 1
```
so it looks like there's already work to do elsewhere to make SWA work with sharded models.

Reply:
Yes, it isn't supported and might be complex to do so.

I believe it would safer to raise a MisConfigurationExecption if this Callback is used with DeepSpeed, DDPSharded, or FSDP.

Reply:
OK thanks, I've added a check for these now and raise a MisconfigurationException